src/persist-client/src/iter.rs

```diff
@@ -0,0 +1,700 @@
+                *self = Self::from_encoded(handle.await??, as_of);
+            }
+            ConsolidationPart::Encoded { .. } | ConsolidationPart::Sorted { .. } => {}
+        }
+        Ok(())
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        match self {
+            ConsolidationPart::Encoded { part, cursor, .. } => cursor.clone().peek(part).is_none(),
+            ConsolidationPart::Sorted { data, index } => data.len() <= *index,
+            ConsolidationPart::Queued { .. } | ConsolidationPart::Prefetched { .. } => false,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Consolidator<T, D> {
+    runs: Vec<VecDeque<(ConsolidationPart<T, D>, usize)>>,
+    as_of: Antichain<T>,
+    budget: usize,
+    // The initial state and the drop stash are used to "carry over" state between calls to
+    // [Self::next], so any "in progress" updates are not lost.
+    initial_state: Option<Tuple<T, D>>,
+    drop_stash: Option<Tuple<T, D>>,
+}
+
+impl<T: Timestamp + Codec64 + Lattice, D: Codec64 + Semigroup> Consolidator<T, D> {
+    pub fn new(as_of: Antichain<T>, prefetch_budget_bytes: usize) -> Self {
+        Self {
+            runs: vec![],
+            as_of,
+            budget: prefetch_budget_bytes,
+            initial_state: None,
+            drop_stash: None,
+        }
+    }
+
+    pub fn enqueue_run<'a>(
+        &mut self,
+        shard_id: ShardId,
+        blob: &Arc<dyn Blob + Send + Sync>,
+        metrics: &Arc<Metrics>,
+        read_metrics: fn(&BatchPartReadMetrics) -> &ReadMetrics,
+        shard_metrics: &Arc<ShardMetrics>,
+        desc: &Description<T>,
+        parts: impl IntoIterator<Item = &'a HollowBatchPart>,
+    ) {
+        let run = parts
+            .into_iter()
+            .map(|part: &HollowBatchPart| {
+                let c_part = ConsolidationPart::Queued {
+                    data: Some(FetchData {
+                        shard_id,
+                        blob: Arc::clone(blob),
+                        metrics: Arc::clone(metrics),
+                        read_metrics,
+                        shard_metrics: Arc::clone(shard_metrics),
+                        part_key: part.key.clone(),
+                        part_desc: desc.clone(),
+                    }),
+                };
+                let size = part.encoded_size_bytes;
+                (c_part, size)
+            })
+            .collect();
+        self.runs.push(run);
+    }
+
+    /// Tidy up: discard any empty parts, and discard any runs that have no parts left.
+    fn trim(&mut self) {
+        self.runs.retain_mut(|run| {
+            while run.front().map_or(false, |(part, _)| part.is_empty()) {
+                run.pop_front();
+            }
+            !run.is_empty()
+        });
+
+        // Some budget may have just been freed up: start prefetching.
+        self.start_prefetches();
+    }
+
+    fn iter(&mut self) -> Option<ConsolidatingIter<T, D>> {
+        self.trim();
```

Comment:
> I think it would be possible to skip cloning the cursor in is_empty if we make is_empty take &mut self (which is a bit weird but might be okay)

This worked out!

> I don't have a good enough grasp yet to know whether anything else will advance the cursor before the next trim

IIUC it should, though that's still throwaway work on the first call, and anyways the reasoning was pretty subtle. It seems much more straightforward to just never clone the cursor so progress is not lost and reasoning is simpler.